### ISSUE TITLE
rocRAND crashes when compiled with avx2 support

### ISSUE THREAD

**Issue by AngryLoki** [ROCm/rocRAND#403]
**Describe the bug**
Due to unaligned allocations in heap with `new` of structs that have `__attribute__((ext_vector_type(4)))` inside them (from amd_hip_vector_types.h and other places) rocRAND crashes when compiled with AVX2 support.

**To Reproduce**
1. Install rocRAND 5.7.1
2. Build with `-march=znver4` (optimization level does not matter)
3. Run tests, e. g. test_rocrand_generate
4. It crashes at line https://github.com/ROCmSoftwarePlatform/rocRAND/blob/rocm-5.7.1/library/src/rng/threefry4x64_20.hpp#L219

**Tech details**

rocRAND 5.7.1 allocates generators in `rocrand_create_generator` with generic `new`, which generally has 128-bit alignment on 64-bit systems. However due to usage of ext_vector_type, for methods of generator clang-17 generates code that expect 256-bit aligned structures. And while inside structures everything is fine, due to heap allocation the whole structure is randomly 128-bit aligned, which causes crash.

```
vmovdqu %xmm1,0x48(%rdi)
vmovdqa %ymm0,0x60(%rdi) <- attempt to copy aligned structures, SIGSEGV
```

**Environment**
HIP version: 5.7.31921
clang version 17.0.2
custom build for Gentoo

**Additional context**
In Eigen similar problem was solved in 2 ways: for C++17-compatible compilers there is nothing to do, for older compilers overloaded new/delete are required.
https://github.com/search?q=repo%3Aeigen-mirror%2Feigen%20EIGEN_HAS_CXX17_OVERALIGN&type=code

Patch in Gentoo: https://github.com/gentoo/gentoo/pull/33400/commits/54e4e7d5ebb3604954edba89d72d4252af249778#diff-e3383e414e6cc5ffd2e0222a1d4152298699139ea2dc379238423f31971149ae

**Comment by Snektron** [comment]
Thanks for the detailed report. I think that a the library itself may be built in C++17 mode soon, so that should be an easy fix. Do you know why exactly the problem is resolved with C++17?

**Comment by AngryLoki** [comment]
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0035r4.html describes the change in C++17.

To put it very simply, for modules compiled in C++17 mode, if structures have aligned fields, the compiler will choose the more appropriate overloaded new / delete operator.

**Comment by Naraenda** [comment]
This was merged with #404
